In [1]:
#PLEASE VIEW AT:
#https://kokes.github.io/nbviewer.js/viewer.html#aHR0cHM6Ly9naXRodWIuY29tL04tQS1NTC9EYXRhX0dsYWNpZXJfRmluYWxfUHJvamVjdF9XZWVrXzdfdG9fMTMvYmxvYi9tYWluL1dlZWslMjAxMi9XZWVrXzEyX25vdGVib29rLmlweW5i

<font color="blue" size="5px"> This is the modelling component of the final project for the internship with Data Glacier. We will explore various models. ROC-AUC will be used as the main scoring metric, since the classes in the target are somewhat imbalanced </font>. 

<font color="red" size="5px"> Let's begin by importing each dataset, shuffling and defining new dataframes: </font>

Import, shuffle and reset index:

In [2]:
import pandas as pd
import numpy as np
df_modified=pd.read_csv(r'df_modified.csv', engine='python').sample(frac=1).reset_index(drop=True) #cleaned dataset, with some rows (with outliers) removed
df_modified2= pd.read_csv(r'df_modified_all_rows.csv', engine='python').sample(frac=1).reset_index(drop=True) #cleaned dataset but without rows (with outliers) removed - has been created for the modelling.

In [3]:
print(df_modified.shape,df_modified2.shape)

(1215, 115) (3424, 115)


Let's create a dataframes with the 6 best features (found in weeks 8-9) and the target variable:

In [4]:
data_best = [df_modified["Dexa_During_Rx_Y"] , df_modified["Comorb_Encounter_For_Screening_For_Malignant_Neoplasms_Y"], df_modified["Comorb_Encounter_For_Immunization_Y"] , df_modified["Comorb_Encntr_For_General_Exam_W_O_Complaint,_Susp_Or_Reprtd_Dx_Y"] , df_modified["Comorb_Long_Term_Current_Drug_Therapy_Y"] , df_modified["Concom_Viral_Vaccines_Y"], df_modified["Persistency_Flag_Persistent"]]
headers = ["Dexa_During_Rx_Y" , "Comorb_Encounter_For_Screening_For_Malignant_Neoplasms_Y", "Comorb_Encounter_For_Immunization_Y" ,"Comorb_Encntr_For_General_Exam_W_O_Complaint,_Susp_Or_Reprtd_Dx_Y" , "Comorb_Long_Term_Current_Drug_Therapy_Y" , "Concom_Viral_Vaccines_Y", "Persistency_Flag_Persistent"]
df_best=pd.concat(data_best, axis=1, keys=headers)

data_best2 = [df_modified2["Dexa_During_Rx_Y"] , df_modified2["Comorb_Encounter_For_Screening_For_Malignant_Neoplasms_Y"], df_modified2["Comorb_Encounter_For_Immunization_Y"] , df_modified2["Comorb_Encntr_For_General_Exam_W_O_Complaint,_Susp_Or_Reprtd_Dx_Y"] , df_modified2["Comorb_Long_Term_Current_Drug_Therapy_Y"] , df_modified2["Concom_Viral_Vaccines_Y"], df_modified2["Persistency_Flag_Persistent"]]
headers = ["Dexa_During_Rx_Y" , "Comorb_Encounter_For_Screening_For_Malignant_Neoplasms_Y", "Comorb_Encounter_For_Immunization_Y" ,"Comorb_Encntr_For_General_Exam_W_O_Complaint,_Susp_Or_Reprtd_Dx_Y" , "Comorb_Long_Term_Current_Drug_Therapy_Y" , "Concom_Viral_Vaccines_Y", "Persistency_Flag_Persistent"]
df_best2=pd.concat(data_best2, axis=1, keys=headers)

Let's create a way to quickly evaluate a model:

In [5]:
from sklearn.metrics import accuracy_score, f1_score, precision_score, recall_score, roc_auc_score
from sklearn.model_selection import cross_val_score

def evaluate():
    print('Precision score is {}'.format(precision_score(y_test,preds, average='macro'))) #use macro scoring since classes are imbalanced
    print('Recall score is {}'.format(recall_score(y_test,preds, average='macro'))) #use macro scoring since classes are imbalanced
    print('F1 score is {}'.format(f1_score(y_test,preds, average='macro'))) #use macro scoring since classes are imbalanced
    print('Accuracy score is {}'.format(accuracy_score(y_test,preds))) 
    print('ROC_AUC score is {}'.format(roc_auc_score(y_test,preds))) 


<font color="red" size="5px"> We will test each dataset on a basic logistic regression model and choose the best dataset for upcoming modelling </font>

First let's try using the cleaned data, without outliers removed, and with all of the features with a basic model:

In [6]:
from sklearn.model_selection import train_test_split
#let the feature dataframe contain every column of df, except the value we are predicting,Persistency_Flag_Persistent
X=df_modified2.loc[:,df_modified2.columns!="Persistency_Flag_Persistent"]
#let the target array contain only the value we are predicting, Persistency_Flag_Persistent
# y=df.loc[:,df.columns=="heart_disease"].values.ravel()
y=df_modified2.loc[:,df_modified2.columns=="Persistency_Flag_Persistent"].values.ravel()
X_train, X_test, y_train, y_test=train_test_split(X,y,test_size=0.30,random_state=123)  


Let's create an evaluate a simple logistic regression model:

In [7]:
from sklearn.linear_model import LogisticRegression
logreg=LogisticRegression(max_iter=1000000)
logreg.fit(X_train,y_train)
preds=logreg.predict(X_test)
evaluate()

Precision score is 0.8273826985797856
Recall score is 0.8043816583237069
F1 score is 0.8126224652540442
Accuracy score is 0.8287937743190662
ROC_AUC score is 0.8043816583237068


Let's try again, but with outliers removed and still using all of the features:

In [8]:
X=df_modified.loc[:,df_modified.columns!="Persistency_Flag_Persistent"]
y=df_modified.loc[:,df_modified.columns=="Persistency_Flag_Persistent"].values.ravel()
X_train, X_test, y_train, y_test=train_test_split(X,y,test_size=0.30,random_state=123)   
logreg=LogisticRegression(max_iter=1000000)
logreg.fit(X_train,y_train)
preds=logreg.predict(X_test)
evaluate()

Precision score is 0.8072344322344323
Recall score is 0.7810701537116631
F1 score is 0.7920938710412394
Accuracy score is 0.8356164383561644
ROC_AUC score is 0.7810701537116631


We used SelectKBest in weeks 8-9 to find the best 6 features. Let's test the dataframe with just the 6 best features and target, and outliers removed:

In [9]:
X=df_best.loc[:,df_best.columns!="Persistency_Flag_Persistent"]
y=df_best.loc[:,df_best.columns=="Persistency_Flag_Persistent"].values.ravel()
X_train, X_test, y_train, y_test=train_test_split(X,y,test_size=0.3,random_state=123)   
logreg=LogisticRegression(max_iter=1000000)
logreg.fit(X_train,y_train)
preds=logreg.predict(X_test)
evaluate()

Precision score is 0.8224592739601422
Recall score is 0.7840751803015954
F1 score is 0.7992589805192926
Accuracy score is 0.8438356164383561
ROC_AUC score is 0.7840751803015954


Let's test the dataframe with just the 6 best features and target, and outliers included:

In [10]:
X=df_best2.loc[:,df_best2.columns!="Persistency_Flag_Persistent"]
y=df_best2.loc[:,df_best2.columns=="Persistency_Flag_Persistent"].values.ravel()
X_train, X_test, y_train, y_test=train_test_split(X,y,test_size=0.30,random_state=123) 

logreg.fit(X_train,y_train)
preds=logreg.predict(X_test)
evaluate()

Precision score is 0.7922754619681993
Recall score is 0.7618887177512514
F1 score is 0.7710294513179128
Accuracy score is 0.7937743190661478
ROC_AUC score is 0.7618887177512514


The df_best dataset seemed to be the best overall to me, based on the evaluation metrics (e.g., high accuracy), and also, there are only 6 features and lower number of rows, so using this data would be computationally efficient, especially since we will be doing some grid searching.We will use this dataset in all future modelling. Also, only needing 6 features to predict persistency may be very useful for the company.

In [11]:
df_best.head()

Dexa_During_Rx_Y  Comorb_Encounter_For_Screening_For_Malignant_Neoplasms_Y  \
0               0.0                                                1.0          
1               0.0                                                0.0          
2               0.0                                                1.0          
3               1.0                                                1.0          
4               0.0                                                0.0          

   Comorb_Encounter_For_Immunization_Y  \
0                                  1.0   
1                                  1.0   
2                                  1.0   
3                                  1.0   
4                                  0.0   

   Comorb_Encntr_For_General_Exam_W_O_Complaint,_Susp_Or_Reprtd_Dx_Y  \
0                                                0.0                   
1                                                0.0                   
2                                                1.0                   
3                                                1.0                   
4                                                1.0                   

   Comorb_Long_Term_Current_Drug_Therapy_Y  Concom_Viral_Vaccines_Y  \
0                                      1.0                      0.0   
1                                      0.0                      0.0   
2                                      0.0                      1.0   
3                                      1.0                      0.0   
4                                      0.0                      0.0   

   Persistency_Flag_Persistent  
0                          1.0  
1                          0.0  
2                          1.0  
3                          1.0  
4                          0.0

<font color="red" size="5px"> Now that we have chosen the best dataset to use, let's try to optimise our logistic regression model    </font>

Let's check the base model's cross-val scores:

In [12]:
X=df_best.loc[:,df_best.columns!="Persistency_Flag_Persistent"]
y=df_best.loc[:,df_best.columns=="Persistency_Flag_Persistent"].values.ravel()
X_train, X_test, y_train, y_test=train_test_split(X,y,test_size=0.3,random_state=123)  

logreg=LogisticRegression(max_iter=1000000)
scores = cross_val_score(logreg, X, y, cv=10, scoring='roc_auc')#
print('Cross-validation scores are: {}'.format(scores))
scores_mean=np.mean(scores)
print('The mean cross-validation score is:{}'.format(scores_mean))

Cross-validation scores are: [0.82158521 0.86661367 0.88060413 0.85325914 0.87583466 0.90267053
 0.87532175 0.86277349 0.86695624 0.80630631]
The mean cross-validation score is:0.8611925132784266


It looks like the base model is good at generalising to new data.

Let's optimise the logistic regression model on df_best (uncomment to run):

In [13]:
# from sklearn.model_selection import RandomizedSearchCV, GridSearchCV
# from sklearn.model_selection import StratifiedKFold

# logreg = LogisticRegression(max_iter=1000000,
#                      n_jobs=-1)

# params={"C":[0.001, 0.01, 0.1, 1, 10, 100, 1000] }




# skf = StratifiedKFold(n_splits=10, shuffle = False) # make sure class balance in each fold is the same as in the orginal dataset

# grid_search=GridSearchCV(logreg,params,cv=skf.split(X_train,y_train),scoring='roc_auc')


# grid_search.fit(X_train, y_train)

# #print(grid_search.cv_results_)
# print(grid_search.best_score_, grid_search.best_estimator_)


We now have a logistic regression model with optimal parameters:

In [14]:
logreg_new=LogisticRegression(C=1, max_iter=1000000, n_jobs=-1)

Let's use the best parameters (i.e., the ones yielding the highest average cross-validation score, 0.867) from grid search:

In [15]:
logreg_new=LogisticRegression(C=1, max_iter=1000000, n_jobs=-1).fit(X_train,y_train)
preds=logreg_new.predict(X_test)
evaluate()

Precision score is 0.8224592739601422
Recall score is 0.7840751803015954
F1 score is 0.7992589805192926
Accuracy score is 0.8438356164383561
ROC_AUC score is 0.7840751803015954


<font color="red" size="5px"> Let's create an XGB classifier and find the best hyperparameters: </font>

Let's create an XGB Classifier with mostly default paramters:

In [16]:
from xgboost import XGBClassifier
xgb=XGBClassifier(eval_metric='mlogloss',use_label_encoder=False)
xgb.fit(X_train,y_train)

#check cross val scores:

scores = cross_val_score(xgb, X, y, cv=10, scoring='roc_auc')#
print('Cross-validation scores are: {}'.format(scores))
scores_mean=np.mean(scores)
print('The mean cross-validation score is:{}'.format(scores_mean))

preds=xgb.predict(X_test)
evaluate()

Cross-validation scores are: [0.83192356 0.87424483 0.87138315 0.84499205 0.86502385 0.87853925
 0.86792149 0.84572072 0.84089447 0.82432432]
The mean cross-validation score is:0.8544967695470792
Precision score is 0.8272484787863632
Recall score is 0.7860056822320973
F1 score is 0.8021220570012392
Accuracy score is 0.8465753424657534
ROC_AUC score is 0.7860056822320973


We ran grid searches with various parameters ranges and attempted to find the best set of parameters (uncomment to run):

In [17]:
# from sklearn.model_selection import GridSearchCV
# from sklearn.model_selection import StratifiedKFold

# xgb = XGBClassifier(#learning_rate=0.03, 
#     #n_estimators=1000, objective='binary:logistic',
#                      n_jobs=-1, use_label_encoder=False)

# params = {
#              'learning_rate':[0.3],
#              'n_estimators':[50,100], 
#              'min_child_weight': [1, 5, 10],
#              'gamma': [0, 1, 2],
#              'subsample': [0.6, 0.8, 1.0],
#              'colsample_bytree': [0.6, 0.8, 1.0],
#              'max_depth': [3,4]
#         }



# skf = StratifiedKFold(n_splits=10, shuffle = False)

# grid_search = GridSearchCV(xgb, param_grid=params,  
#                              scoring='roc_auc', n_jobs=-1, cv=skf.split(X_train,y_train), verbose=3
#                             )
# grid_search.fit(X_train, y_train)

# #print(grid_search.cv_results_)

# print(grid_search.best_score_, grid_search.best_estimator_)

We shall use the best parameters (those yielding the highest average cross val score(0.870)) from grid search:

In [18]:
xgb_new=XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=0.8, gamma=2, gpu_id=-1,
              importance_type='gain', interaction_constraints='',
              learning_rate=0.3, max_delta_step=0, max_depth=4,
              min_child_weight=5, #missing=nan,
              monotone_constraints='()',
              n_estimators=100, n_jobs=-1, num_parallel_tree=1, random_state=0,
              reg_alpha=0, reg_lambda=1, scale_pos_weight=1, subsample=0.6,
              tree_method='exact', use_label_encoder=False,
              validate_parameters=1, verbosity=None,
                     eval_metric='mlogloss' #specify eval metric to avoid warnings 
                     )

xgb_new.fit(X_train,y_train)

#test on test data now:
preds=xgb_new.predict(X_test)
evaluate()

Precision score is 0.8206114639309559
Recall score is 0.786861659503169
F1 score is 0.8005751258087707
Accuracy score is 0.8438356164383561
ROC_AUC score is 0.786861659503169


<font color="red" size="5px">  Let's create a neural network and optimise its hyperparameters: </font>

Import necessary packages:

In [19]:
from keras.models import Sequential
from keras.layers import Dense
from keras import metrics
import tensorflow
from keras.wrappers.scikit_learn import KerasClassifier

Define the network architecture and wrapper for KerasClassifier:

In [20]:
def create_nn(optimizer='uniform', init='adam'):
    nn = Sequential()
    nn.add(Dense(5, input_dim=X_train.shape[1], activation='relu')) #let's use 2/3 size of input layer + size of output layer for number of nodes
    nn.add(Dense(5, activation='relu'))
    nn.add(Dense(1, activation='relu'))
    nn.compile(loss='binary_crossentropy', optimizer='adam', 
               #metrics='roc-auc'
              )
    return nn

In [21]:
nn=KerasClassifier(build_fn=create_nn, verbose=0)

We used grid search to search for optimal parameters (we used various ranges and narrowed the parameters down). Uncomment to run:

In [22]:
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import StratifiedKFold
optimizers = [#'rmsprop',
    'adam'
]
init = [#'glorot_uniform'
    #'normal',
         'uniform'
]
epochs = [150]
batches = [5,10]

params = dict(optimizer=optimizers, epochs=epochs, 
                  batch_size=batches,
                  init=init)

skf = StratifiedKFold(n_splits=10, shuffle = False)

grid_search = GridSearchCV(estimator=nn, param_grid=params, n_jobs=-1, cv=skf.split(X_train,y_train))
grid_search = grid_search.fit(X_train.values, y_train)

C:\Users\Natha\anaconda3\lib\site-packages\sklearn\model_selection\_search.py:918: UserWarning: One or more of the test scores are non-finite: [nan nan]
  warnings.warn(


Let's use the params which acheived the best average cross-validation score:

In [23]:
nn_new=KerasClassifier(build_fn=create_nn, verbose=0,batch_size=5, epochs=150, init='uniform', optimizer='adam')
nn_new._estimator_type="classifier"

Train the model:

In [24]:
nn_new.fit(X_train,y_train)

Predict and evaluate:

In [27]:
preds = nn_new.predict(X_test)
evaluate()

C:\Users\Natha\AppData\Roaming\Python\Python38\site-packages\keras\engine\sequential.py:450: UserWarning: `model.predict_classes()` is deprecated and will be removed after 2021-01-01. Please use instead:* `np.argmax(model.predict(x), axis=-1)`,   if your model does multi-class classification   (e.g. if it uses a `softmax` last-layer activation).* `(model.predict(x) > 0.5).astype("int32")`,   if your model does binary classification   (e.g. if it uses a `sigmoid` last-layer activation).
  warnings.warn('`model.predict_classes()` is deprecated and '


Precision score is 0.8348545469700759
Recall score is 0.7926531652946747
F1 score is 0.809189126394052
Accuracy score is 0.852054794520548
ROC_AUC score is 0.7926531652946748


Best: 0.793230 using {'batch_size': 20, 'epochs': 100, 'init': 'uniform', 'optimizer': 'adam'}

<font color="red" size="5px"> Let's create a random forest and find the best hyperparameters: </font>



Let's create a random forest classifier with default settings and check the cross val scores:

In [28]:
from sklearn.ensemble import RandomForestClassifier

rf=RandomForestClassifier()

scores = cross_val_score(rf, X, y, cv=10, scoring='roc_auc')#
print('Cross-validation scores are: {}'.format(scores))
scores_mean=np.mean(scores)
print('The mean cross-validation score is:{}'.format(scores_mean))

Cross-validation scores are: [0.83599624 0.86502385 0.87965024 0.84689984 0.86391097 0.876287
 0.86888674 0.84765122 0.83880309 0.82496782]
The mean cross-validation score is:0.8548077018858754


Train, predict and evaluate:

In [29]:
rf.fit(X_train,y_train)
preds=rf.predict(X_test)
evaluate()

Precision score is 0.8348545469700759
Recall score is 0.7926531652946747
F1 score is 0.809189126394052
Accuracy score is 0.852054794520548
ROC_AUC score is 0.7926531652946748


In [30]:
# # from sklearn.model_selection import GridSearchCV
# # from sklearn.model_selection import StratifiedKFold


# params= {'bootstrap': [True],
#  'max_depth': [11],
#  'max_features': ['auto'],
#  'min_samples_leaf': [7,8],
#  'min_samples_split': [9,10],
#  'n_estimators': [1000]}

# skf = StratifiedKFold(n_splits=10, shuffle = False)
# grid_search = GridSearchCV(estimator=rf, param_grid=params, n_jobs=-1, scoring='roc_auc' ,cv=skf.split(X_train,y_train))
# grid_search = grid_search.fit(X_train, y_train)

# print(grid_search.best_score_,grid_search.best_params_)
# #print(grid_search.cv_results_)
# #grid_search.best_estimator_

The best average cross-val score was 0.864. Let's use the estimator which achieved this score:

In [31]:
rf_new=RandomForestClassifier(**{'bootstrap': True, 'max_depth': 11, 'max_features': 'auto', 'min_samples_leaf': 8, 'min_samples_split': 9, 'n_estimators': 1000})


Train, predict and evaluate:

In [32]:
rf_new.fit(X_train,y_train)
preds=rf_new.predict(X_test)
evaluate()

Precision score is 0.8168067226890756
Recall score is 0.7746412180374445
F1 score is 0.7908187549174834
Accuracy score is 0.8383561643835616
ROC_AUC score is 0.7746412180374445


<font color="red" size="5px">Let's create a K nearest neighbors classifier and find the best value for K: </font>

Let's define a KNN with default parameters, and check the cross-validation scores:

In [33]:
from sklearn.neighbors import  KNeighborsClassifier
knn=KNeighborsClassifier()

scores = cross_val_score(knn, X, y, cv=10, scoring='roc_auc')#
print('Cross-validation scores are: {}'.format(scores))
scores_mean=np.mean(scores)
print('The mean cross-validation score is:{}'.format(scores_mean))

Cross-validation scores are: [0.82095865 0.83084261 0.86152623 0.78314785 0.77281399 0.85183398
 0.81418919 0.79697555 0.76994852 0.79584942]
The mean cross-validation score is:0.8098085984037997


cross-val scores are quite close together, except for the 0.63 score.

let's train the knn and predict and evaluate on the test data:

In [34]:
knn.fit(X_train,y_train)
preds=knn.predict(X_test)
acc_score=accuracy_score(y_test,preds)
evaluate()

Precision score is 0.8316367177126671
Recall score is 0.7729292634953012
F1 score is 0.7936523679642944
Accuracy score is 0.8438356164383561
ROC_AUC score is 0.7729292634953011


Let's create a grid search to find the best value of K (uncomment to run)

In [35]:
# # from sklearn.model_selection import GridSearchCV
# # from sklearn.model_selection import StratifiedKFold

# params = {
#     'n_neighbors':n_neigh
# }

# # skf = StratifiedKFold(n_splits=10, shuffle = False)
#n_neigh=[i for i in range(1,33) if i%2!=0]
# grid_search = GridSearchCV(estimator=knn, param_grid=params, n_jobs=-1, scoring='roc_auc' ,cv=skf.split(X_train,y_train))
# grid_search = grid_search.fit(X_train, y_train)

# print(grid_search.best_score_,grid_search.best_params_)
# #print(grid_search.cv_results_)
# #grid_search.best_estimator_

let's use the parameters with the highest average cross-val score (0.853) and predict and evalute on the test data:

In [36]:
knn_new=KNeighborsClassifier(n_neighbors=31)
preds=knn.predict(X_test)
evaluate()

Precision score is 0.8316367177126671
Recall score is 0.7729292634953012
F1 score is 0.7936523679642944
Accuracy score is 0.8438356164383561
ROC_AUC score is 0.7729292634953011


<font color="red" size="5px">Let's use stacking: </font>

Let's use the top 3 estimators (from the code blocks above):

In [37]:
xgb_new=XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=0.8, gamma=2, gpu_id=-1,
              importance_type='gain', interaction_constraints='',
              learning_rate=0.3, max_delta_step=0, max_depth=4,
              min_child_weight=5, #missing=nan,
              monotone_constraints='()',
              n_estimators=100, n_jobs=-1, num_parallel_tree=1, random_state=0,
              reg_alpha=0, reg_lambda=1, scale_pos_weight=1, subsample=0.6,
              tree_method='exact', use_label_encoder=False,
              validate_parameters=1, verbosity=None,
                     eval_metric='mlogloss' #specify eval metric to avoid warnings 
                     )

nn_new=KerasClassifier(build_fn=create_nn, verbose=0,batch_size=5, epochs=150, init='uniform', optimizer='adam')
nn_new._estimator_type="classifier"

rf_new=RandomForestClassifier(**{'bootstrap': True, 'max_depth': 11, 'max_features': 'auto', 'min_samples_leaf': 8, 'min_samples_split': 9, 'n_estimators': 1000})

Define the splitting method and declare the estimators:

In [38]:
from sklearn.ensemble import StackingClassifier
from sklearn.model_selection import StratifiedKFold

skf = StratifiedKFold(n_splits=10, shuffle = False)
estimators=[('xgb_new', xgb_new), ('nn_new', nn_new), ('rf_new',rf_new)]

Define the stacking model:

In [39]:
sc = StackingClassifier(
    estimators=estimators,
    cv=skf.split(X_train,y_train))    

Fit to the training data:

In [40]:
sc.fit(X_train,y_train)

StackingClassifier(cv=<generator object _BaseKFold.split at 0x0000016A9E55BEB0>,
                   estimators=[('xgb_new',
                                XGBClassifier(base_score=0.5, booster='gbtree',
                                              colsample_bylevel=1,
                                              colsample_bynode=1,
                                              colsample_bytree=0.8,
                                              eval_metric='mlogloss', gamma=2,
                                              gpu_id=-1, importance_type='gain',
                                              interaction_constraints='',
                                              learning_rate=0.3,
                                              max_delta_step=0, max_depth=4,
                                              min_chi...
                                              num_parallel_tree=1,
                                              random_state=0, reg_alpha=0,
                    

Predict and evaluate:

In [41]:
preds=sc.predict(X_test)
evaluate()

Precision score is 0.8177715786019035
Recall score is 0.7821446783710935
F1 score is 0.7964109863055855
Accuracy score is 0.8410958904109589
ROC_AUC score is 0.7821446783710934


<font color="red" size="5px">Let's use voting: </font>

Let's define the voting classifier model. The neural network performed well so we will give it a relatively high weight:

In [42]:
from sklearn.ensemble import VotingClassifier
vc = VotingClassifier(estimators= [('xgb_new', xgb_new),
                                   ('nn_new', nn_new), 
                                   ('rf_new',rf_new)], 
                      voting='soft', #vote based on probabilities rather than majority vote of classes
                      weights=[1,2,1] #assign weights: xgboost performed very well so we'll give it a relatively high weight
                     )

Train the model:

In [43]:
vc.fit(X_train,y_train)

VotingClassifier(estimators=[('xgb_new',
                              XGBClassifier(base_score=0.5, booster='gbtree',
                                            colsample_bylevel=1,
                                            colsample_bynode=1,
                                            colsample_bytree=0.8,
                                            eval_metric='mlogloss', gamma=2,
                                            gpu_id=-1, importance_type='gain',
                                            interaction_constraints='',
                                            learning_rate=0.3, max_delta_step=0,
                                            max_depth=4, min_child_weight=5,
                                            missing=nan,
                                            monotone_constraints='()',
                                            n_estimat...
                                            reg_alpha=0, reg_lambda=1,
                                            

Predict and evaluate:

In [44]:
preds=vc.predict(X_test)
evaluate()

Precision score is 0.3547945205479452
Recall score is 0.5
F1 score is 0.41506410256410253
Accuracy score is 0.7095890410958904
ROC_AUC score is 0.5


C:\Users\Natha\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


<font color="green" size="5px">Which model is best? </font>

The XGB classifier (xgb_new) was definitely a good model, with solid scores overall, and some relatively high precision and recall scores. The default model for xgb boost had a good cross-validation score and this wasn't found to change significantly after parameter tuning.  The neural network (nn_new) performed slightly better (or the same) compared to xgb_new in general. The random forest (rf_new) also performed well but was slighly worse than the other models. Both the voting classifer (vc) and stacking classifer (sc) performed well in general but weren't better than nn_new. Overall the best model is nn_new.

These are the evaluation metrics for nn_new on the test data when I ran the code (we found these earlier):    

In [ ]:
print("Precision score 0.8175742386268702")
print("Recall score is 0.777711363485422")
print("F1 score is 0.7911991199119912")
print("Accuracy score is 0.821917808219178")
print("ROC_AUC score is 0.777711363485422")

In [45]:
X_train.shape

(850, 6)

In [46]:
df_best.head()

Dexa_During_Rx_Y  Comorb_Encounter_For_Screening_For_Malignant_Neoplasms_Y  \
0               0.0                                                1.0          
1               0.0                                                0.0          
2               0.0                                                1.0          
3               1.0                                                1.0          
4               0.0                                                0.0          

   Comorb_Encounter_For_Immunization_Y  \
0                                  1.0   
1                                  1.0   
2                                  1.0   
3                                  1.0   
4                                  0.0   

   Comorb_Encntr_For_General_Exam_W_O_Complaint,_Susp_Or_Reprtd_Dx_Y  \
0                                                0.0                   
1                                                0.0                   
2                                                1.0                   
3                                                1.0                   
4                                                1.0                   

   Comorb_Long_Term_Current_Drug_Therapy_Y  Concom_Viral_Vaccines_Y  \
0                                      1.0                      0.0   
1                                      0.0                      0.0   
2                                      0.0                      1.0   
3                                      1.0                      0.0   
4                                      0.0                      0.0   

   Persistency_Flag_Persistent  
0                          1.0  
1                          0.0  
2                          1.0  
3                          1.0  
4                          0.0